cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java

```diff
@@ -154,8 +157,11 @@ public DynamicDatasetContext(TransactionContext context, DatasetFramework datase
         throw new DatasetInstantiationException(String.format("Dataset '%s' does not exist", name));
       }
 
-      if (dataset instanceof TransactionAware) {
+      // For every instance of a TransactionAware dataset acquired, add it to the
+      // current transaction only if it hasn't been added previously.
+      if (!txnInProgressDatasets.contains(datasetKey) && dataset instanceof TransactionAware) {
```

Comment:
Did it this way since if the set contains that key it implicitly implies that a dataset is TxAware, so you avoid the additional check. But I agree readability trumps that.